c/src/point_one/polaris/polaris.c

```diff
@@ -561,81 +647,178 @@ int Polaris_Work(PolarisContext_t* context) {
-    // service.
+    // If the connection was closed by the server, print the reason.
     else {
-      if (context->disconnected) {
-        P1_DebugPrint("Connection terminated by user request.\n");
+      if (bytes_read == 0) {
+        // Note that this is considered unexpected, even though the server
+        // closed the connection without error. Once connected, we expect the
+        // stream to stay open indefinitely under normal circumstances. The
+        // server should not have a reason to terminate the connection.
+        P1_Print("Warning: Connection terminated remotely.\n");
       } else {
-        P1_DebugPrint("Connection terminated upstream.\n");
+#ifdef P1_FREERTOS
+        bytes_read = original_bytes_read;
+        errno = original_errno;
+#endif
+        P1_PrintReadWriteError(
+            context, "Warning: Connection terminated remotely", bytes_read);
       }
-      CloseSocket(context, 1);
-      return POLARIS_CONNECTION_CLOSED;
     }
+
+    // If we connected to the server but never got any data, it was likely:
+    // 1. The user failed to send either a position (that is within range of a
+    //    base station in the network) or a request for a specific base station
+    // 2. The authentication token was rejected and socket was closed remotely
+    //
+    // Note that in general for case (2), we do expect the server to send an
+    // RTCM 1029 message indicating the reason for failure. That would be
+    // handled as "data received but not authenticated" (authenticated == 0),
+    // followed by EINTR or a similar condition above. total_bytes_received will
+    // not be 0.
+    //
+    // Similarly, if we received something from the server but no position or
+    // beacon request has been sent, the data we received is likely an
+    // authentication error. This too should be treated as "data received but
+    // not authenticated" (authenticated == 0).
+    int ret = POLARIS_CONNECTION_CLOSED;
+    if (context->authenticated == POLARIS_AUTHENTICATED) {
+      // Note that currently, we only declare authenticated == 1 after we
+      // receive enough data to confirm we did not get an error 1029 response,
+      // so we do not expect to get here if the user did not send either a
+      // position or base station request.
+      //
+      // No warning needed. If there was a socket error, we'll have printed a
+      // warning above.
+    }
+    // Authentication skipped and data received.
+    else if (context->total_bytes_received > 0 &&
+             context->authenticated == POLARIS_AUTHENTICATION_SKIPPED) {
+      // No warning necessary.
+    }
+    // Not authenticated, but response received: likely server indicated an
+    // error.
+    else if (context->total_bytes_received > 0) {
+      P1_Print(
+          "Warning: Polaris connection closed with an error response. Is "
+          "your authentication token valid?\n");
```

Comment:
Can we print the error?

Reply:
Not without adding an RTCM parser. If we hit an errno it will be printed above before we get here, but if we get an error notification from the server it comes as an RTCM 1029 message. We could add functionality to parse it and verify the CRC, but we don't have that in this library currently.